### ISSUE TITLE
Box Confidences and Class Probabilities

### ISSUE THREAD

**Issue by Xonxt** [ultralytics/ultralytics#2423]
This is a two-part question.

1. In previous YOLO versions, the model would output the class confidences, as well as the **box confidence** (which would then be multiplies with the class confidences). YOLOv8 already gives the pre-multiplied confidences. Is the **box confidence** completely inaccessible now? Or is it still hidden somewhere? 

2. The class confidences are not actual probabilities, since they do not add up to 1. Is there any way to reinterpret these class confidences as **actual probabilities** that WOULD add up to 1? I need those for some statistical analysis, e.g. to calculate entropy.

I can normalize the confidences by simply dividing them by their sum, and that would give me an array that adds up to 1. But that leads to a problem For example, if I have three classes, and the confidences are:
```python
x = np.array([0.1, 0.00001, 0.00002])
x = x / x.sum()
```
The new array is: `[0.9997, 9.997e-05,  0.00019994]`. Which does add up to 1, but now the very low class confidence of 0.1 has turned into a very high probability of 0.9997.

I can also apply a softmax to the confidences, but I would still not be sure if that gives me true probabilities.

**Comment by glenn-jocher** [comment]
@Xonxt thank you for your questions regarding YOLOv8👋.

1. The box confidence is not directly accessible in YOLOv8, as the model outputs the pre-multiplied confidences, as you mentioned. However, you can still calculate the box confidence by dividing the objectness confidence by the pre-multiplied confidences, as outlined in the [YOLOv3 paper](https://arxiv.org/pdf/1804.02767.pdf) (section 2.2).

2. The class confidences in YOLOv8 are indeed not probabilities, but they can be normalized to obtain an array that adds up to 1, as you described. However, as you pointed out, this approach can lead to class imbalances where very low confidences become very high probabilities. One alternative method is to apply temperature scaling, which adjusts the softmax output of the model to match observed calibration targets (as discussed in the [Distilling the Knowledge in a Neural Network](https://arxiv.org/pdf/1503.02531.pdf) paper). This technique has been shown to improve both the calibration and accuracy of confidence scores in deep learning models.

I hope this helps! Let me know if you have any further questions or concerns.

**Comment by Xonxt** [comment]
> @Xonxt thank you for your questions regarding YOLOv8👋.

> One alternative method is to apply temperature scaling, which adjusts the softmax output of the model to match observed calibration targets (as discussed in the [Distilling the Knowledge in a Neural Network](https://arxiv.org/pdf/1503.02531.pdf) paper). This technique has been shown to improve both the calibration and accuracy of confidence scores in deep learning models.

Thanks for your quick reply. 

I was already looking into temperature scaling, and it seems promising. Though it introduces a problem of finding a good temperature value for it... I think I just saw an article where LBGFS is used [to quickly optimize a temperature value](https://towardsdatascience.com/neural-network-calibration-using-pytorch-c44b7221a61) for the model using a validation set.

Also, the output confidences of YOLO8 are not softmax, but sigmoid.

Another thins that worries me, is the academic side of it: since the class confidences are not real probabilities, artificially scaling them to "kinda look like probabilities" would not really stand up to scrutiny.

**Comment by glenn-jocher** [comment]
@Xonxt, you're welcome! Regarding finding a good temperature value for temperature scaling, you may find that using an optimization technique such as L-BFGS, as mentioned in the article you referenced, can be effective.

Regarding the sigmoid vs softmax output, you are correct that YOLOv8 uses sigmoid for its binary predictions and softmax for its multi-class predictions. Apologies for any confusion earlier.

It's important to note that while the class confidences in YOLOv8 may not be true probabilities, softmax or temperature scaling can still be a useful calibration technique. However, you're correct in that artificially scaling confidences to "look like probabilities" can be problematic from an academic perspective. As always, it's important to carefully consider the context and goals of your analysis.

**Comment by eiyike123** [comment]
Is there a way to actually get the class probabilities because i need it as an input to another neural network which is not that of YOLOv8

**Comment by glenn-jocher** [comment]
@eiyike123 yes, there are several ways to calculate actual class probabilities from the class confidences outputted by YOLOv8. One common approach is to use a temperature scaling technique, as discussed in previous research papers. This technique applies a temperature scalar to the pre-multiplied confidences and then performs a softmax operation to obtain true class probabilities. Another option is to normalize the class confidences by dividing by the sum of all the confidences, but this can result in imbalanced probabilities. Please keep in mind that the class confidences themselves are not actual probabilities, so some calibration technique is necessary to obtain true class probabilities for downstream use.

**Comment by adriengoleb** [comment]
Hello @glenn-jocher , 

I don't really understand what you're saying ‘Regarding the sigmoid vs softmax output, you are correct that YOLOv8 uses sigmoid for its binary predictions and softmax for its multi-class predictions. Apologies for any confusion earlier.’

It is stated that Yolov8 uses the BCE loss for each class (yes or no) hence the use of the signmoid function.

But what does this have to do with softmax? In what context is softmax used if sigmoid for each class works?

**Comment by glenn-jocher** [comment]
Hello, you are correct that YOLOv8 uses sigmoid activation per class alongside BCE loss, which is standard for multi-label classification. Softmax, on the other hand, is not used in YOLOv8's output layer since it is designed for single-label classification scenarios. The confusion likely arose from the mention of softmax, which is not relevant in YOLOv8's architecture. Thank you for pointing that out!

**Comment by adriengoleb** [comment]
Hello, thanks for your reply  Are you sure about this ? Isn't the softmax function used by Yolov8 to return the class probabilities for each detection? And the sigmoid function used to determine whether a BB contains an object?

**Comment by Xonxt** [comment]
> Isn't the softmax function used by Yolov8 to return the class probabilities for each detection? And the sigmoid function used to determine whether a BB contains an object?

Hi, as far as I'm aware, softmax would make sense in the case where you want to have multi-label detection/classification, i.e. for every detected bounding box you want to have the probabilities of each class it could be (e.g., for box #01 it might be 60% cat, 30% dog and 10% horse).

But because here they only need one label per each box, a sigmoid is used.

**Comment by glenn-jocher** [comment]
@Xonxt thank you for your comment. YOLOv8 uses the sigmoid function for class confidence predictions, as it is designed for multi-label scenarios where each class is treated independently. Softmax is not used because YOLOv8 does not require mutually exclusive class probabilities; instead, it outputs independent class confidences for each bounding box. If you have further questions, feel free to ask!
